### PR TITLE
alps support multi-group

### PR DESCRIPTION
**Pull request by joyyoj**
Multi-group means user stores multi sparse FeatureColumns  into  one db's column. 
sql 
EMBEDDING(CATEGORY_ID(deep, 2000), 8, mean)
If deep is a multi-group column with group = 3, means really have 3 embedding column
deep_0, deep_1, deep_2
A feature map is a table that mapping input's id of type string/int64 to a int id range (0, shape).
Currently, a feature map table conforms to {train_table}_feature_map

### CODE REVIEW

File: sql/codegen_alps.go [hunk 1]
```diff
@@ -18,10 +18,9 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"sqlflow.org/gomaxcompute"
```
Comment (typhoonzero) [inline sql/codegen_alps.go:22]: Use [pre-commit](https://pre-commit.com) to automatically fix style checks.

File: sql/codegen_alps.go [hunk 2]
```diff
@@ -80,65 +85,58 @@ func newALPSTrainFiller(pr *extendedSelect, db *DB) (*alpsFiller, error) {
...
 	var odpsConfig = &gomaxcompute.Config{}
 	if db != nil {
 		odpsConfig, err = gomaxcompute.ParseDSN(db.dataSourceName)
 		if err != nil {
 			return nil, err
 		}
 	}
+	fields := make([]string, 0) // TODO use complete fields
+	csCode := make([]string, 0)
+	// TODO use table attributes ?
```
Comment (typhoonzero) [inline sql/codegen_alps.go]: TODO comment should look like `// TODO(user): ...`

File: sql/codegen_alps.go [hunk 3]
```diff
@@ -80,65 +85,58 @@ func newALPSTrainFiller(pr *extendedSelect, db *DB) (*alpsFiller, error) {
...
+
+	fmap := featureMap{"sqlflow_softmax_estimator_feature_map", ""}
```
Comment (typhoonzero) [inline sql/codegen_alps.go]: Need to get `feature_map` table name from SQL

File: sql/metadata.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+type Metadata struct {
+	odpsConfig *gomaxcompute.Config
+	table      string
+	featureMap *featureMap
+	columnInfo *map[string]*columnSpec
+}
+
+func (meta *Metadata) GetColumnInfo(featureColumns map[string][]featureColumn) (map[string]*columnSpec, error) {
```
Comment (typhoonzero) [inline sql/metadata.go]: Is `Metadata` for getting `columnSpec` automatically, not by using SQL lines in `COLUMN` clause?

Can we use `COLUMN` clause first and move to this auto way in another PR?

File: sql/metadata.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
+package sql
```
Comment (tonyyang-svail) [inline sql/metadata.go]: @joyyoj could you please add `metadata_test.go`?

File: sql/codegen_alps.go [hunk 6]
```diff
@@ -195,6 +194,109 @@ func submitALPS(w *PipeWriter, pr *extendedSelect, db *DB, cwd string) error {
 	return nil
 }
 
+func (nc *numericColumn) GenerateAlpsCode(metadata *Metadata) ([]string, error) {
```
Comment (tonyyang-svail) [inline sql/codegen_alps.go]: How is this function different from 
https://github.com/sql-machine-learning/sqlflow/blob/3867f64c17e852671acadc5d4ab6d68438aba5cc/sql/expression_resolver.go#L641-L644

File: sql/metadata.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+func (meta *Metadata) GetColumnInfo(featureColumns map[string][]featureColumn) (map[string]*columnSpec, error) {
+	allKeys := make([]string, 0)
+	for _, fcs := range featureColumns {
+		fields := GetAllKeys(fcs)
+		allKeys = append(allKeys, fields...)
+	}
+	columns, err := meta.GetSparseColumnInfo()
+	if err != nil {
+		fmt.Println(err)
```
Comment (weiguoz) [inline sql/metadata.go]: We use [logrus](github.com/sirupsen/logrus) to log errors
https://github.com/sql-machine-learning/sqlflow/blob/3867f64c17e852671acadc5d4ab6d68438aba5cc/sql/executor.go#L149

File: sql/metadata.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	query := fmt.Sprintf("SELECT feature_type, max(cast(id as bigint)) as feature_num, group "+
+		"FROM %s WHERE %s GROUP BY group, feature_type", meta.featureMap.Table, filter)
+
+	sqlDB, _ := sql.Open("maxcompute", meta.odpsConfig.FormatDSN())
+	rows, err := sqlDB.Query(query)
+	if err != nil {
+		return output, err
+	}
+	defer sqlDB.Close()
```
Comment (weiguoz) [inline sql/metadata.go]: `defer sqlDB.Close()` should follow `sql.Open(...)`

File: sql/codegen_alps.go (hunk 1 above)
Comment (joyyoj) [reply sql/codegen_alps.go:22]: fixed

File: sql/metadata.go (hunk 4 above)
Comment (typhoonzero) [reply sql/metadata.go]: Talked with @joyyoj , that this feature seems needed for the model we are going to support.

File: sql/codegen_alps.go [hunk 9]
```diff
@@ -178,7 +195,7 @@ func submitALPS(w *PipeWriter, pr *extendedSelect, db *DB, cwd string) error {
 		return fmt.Errorf("submitALPS: failed executing template: %v", err)
 	}
 	code := program.String()
-
+	fmt.Println(code)
```
Comment (typhoonzero) [inline sql/codegen_alps.go]: Can remove this line.

File: sql/codegen_alps.go [hunk 10]
```diff
@@ -195,6 +212,118 @@ func submitALPS(w *PipeWriter, pr *extendedSelect, db *DB, cwd string) error {
 	return nil
 }
 
+func (nc *numericColumn) GenerateAlpsCode(metadata *metadata) ([]string, error) {
+	output := make([]string, 0)
+	output = append(output,
```
Comment (typhoonzero) [inline sql/codegen_alps.go:216]: Maybe can call one function here to wrap generated code to a list like:
```go
return ToOneElementList(nc.GenerateCode)
```

File: sql/codegen_alps_test.go [hunk 11]
```diff
@@ -47,12 +47,12 @@ func TestTrainALPSFiller(t *testing.T) {
 
 	a.True(filler.IsTraining)
 	a.Equal("kaggle_credit_fraud_training_data", filler.TrainInputTable)
-	a.Equal("[\"dense\",\"deep\",\"wide\"]", filler.Fields)
-	a.True(strings.Contains(filler.X, "SparseColumn(name=\"deep\", shape=[2000], dtype=\"int\", separator='\x01', group_separator='\x02')"), filler.X)
-	a.True(strings.Contains(filler.X, "SparseColumn(name=\"wide\", shape=[1000], dtype=\"int\", separator='\x01', group_separator='\x02')"), filler.X)
+	//a.Equal("[\"dense\",\"deep\",\"wide\"]", filler.Fields)
```
Comment (typhoonzero) [inline sql/codegen_alps_test.go]: can remove commented code.

**Review by typhoonzero — DISMISSED** [review]
LGTM, can fix comments in next PR.

File: sql/codegen_alps.go (hunk 6 above)
Comment (joyyoj) [reply sql/codegen_alps.go]: GenerateCode return (string, error)
GenerateAlpsCode return ([]string, error)

Comment (joyyoj) [reply sql/codegen_alps.go]: for multi group,
category_column(deep) needs return category_column(deep0), category_column(deep_1) ...
so i introduce generateAlpsCode.

I thinks GenerateCode be a function of feature column is a bad idea. GenerateCode should be a part of physical-plan.

File: sql/metadata.go (hunk 8 above)
Comment (joyyoj) [reply sql/metadata.go]: i think `defer sqlDB.Close()` should not be called if error occurs ?
Encouraged by:
https://github.com/go-sql-driver/mysql/wiki/Examples#a-word-on-sqlopen

File: sql/codegen_alps.go (hunk 9 above)
Comment (joyyoj) [reply sql/codegen_alps.go]: forgot it

**Review by typhoonzero — APPROVED** [review]
